### PR TITLE
Source Twilio: Migrate to Low Code CDK

### PR DESCRIPTION
**Pull request by aazam-gh**
## What
Migrating source to Low Code CDK
Merges https://github.com/airbytehq/DeveloperWeek-Connector-Contest/issues/19

# How
Developed using the Configuration Based Source

## Recommended reading order
1. `spec.yaml`
2. `mainfest.yaml`
3. `schemas`

## 🚨 User Impact 🚨
No breaking changes. Just migrating entire connector functionality

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- Unit & integration tests added

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- Create a non-forked branch based on this PR and test the below items on it
- Build is successful
- If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).

</details>

### CODE REVIEW

File: airbyte-integrations/connectors/source-twilio/unit_tests/test_streams.py [hunk 1]
```diff
@@ -1,327 +0,0 @@
...
-    UsageRecords,
-    UsageTriggers,
-)
-
-TEST_CONFIG = {
-    "account_sid": "airbyte.io",
-    "auth_token": "secret",
-    "start_date": "2022-01-01T00:00:00Z",
-    "lookback_window": 0,
-}
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-twilio/unit_tests/test_streams.py:34]: We must keep the same behavior to the new connector and not delete all unit tests.

**Review by marcosmarxm — CHANGES_REQUESTED** [review]

File: airbyte-integrations/connectors/source-twilio/Dockerfile [hunk 2]
```diff
@@ -1,16 +1,38 @@
...
+
+# copy payload code only
 COPY main.py ./
+COPY source_twilio ./source_twilio
 
 ENV AIRBYTE_ENTRYPOINT "python /airbyte/integration_code/main.py"
 ENTRYPOINT ["python", "/airbyte/integration_code/main.py"]
 
-LABEL io.airbyte.version=0.5.0
+LABEL io.airbyte.version=0.1.0
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-twilio/Dockerfile:37]: `1.0.0`

**Review by marcosmarxm — CHANGES_REQUESTED** [review]

File: airbyte-integrations/connectors/source-twilio/unit_tests/test_streams.py (hunk 1 above)
Comment (Alcadeus0) [reply airbyte-integrations/connectors/source-twilio/unit_tests/test_streams.py:34]: gotcha

File: airbyte-integrations/connectors/source-twilio/source_twilio/manifest.yaml [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+      datetime_format: "%Y-%m-%d %H:%M:%S.%f+00:00"
+    step: "P1D"
+    datetime_format: "%Y-%m-%d"
+    cursor_granularity: "P1D"
+    cursor_field: "date"
+  retriever:
+    record_selector:
+      $ref: "#/definitions/selector"
+    paginator:
+      type: NoPagination
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-twilio/source_twilio/manifest.yaml:30]: @Alcadeus0 are you planning to add the paginator?

**Review by marcosmarxm — CHANGES_REQUESTED** [review]

File: airbyte-integrations/connectors/source-twilio/source_twilio/manifest.yaml (hunk 3 above)
Comment (Alcadeus0) [reply airbyte-integrations/connectors/source-twilio/source_twilio/manifest.yaml]: @marcosmarxm Yep I have added it

File: airbyte-integrations/connectors/source-twilio/Dockerfile [hunk 4]
```diff
@@ -1,16 +1,38 @@
...
+
+# copy payload code only
 COPY main.py ./
+COPY source_twilio ./source_twilio
 
 ENV AIRBYTE_ENTRYPOINT "python /airbyte/integration_code/main.py"
 ENTRYPOINT ["python", "/airbyte/integration_code/main.py"]
 
-LABEL io.airbyte.version=0.5.0
+LABEL io.airbyte.version=1.0.0
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-twilio/Dockerfile:37]: @erohmensing in some sense moving a python cdk to low-code won't change the behavior of the connector and it isn't a breaking change. I asked to be a major update to make clear about the drastic change into the code. Can you give your opinion here?

Comment (Alcadeus0) [reply airbyte-integrations/connectors/source-twilio/Dockerfile:37]: @marcosmarxm can you take a look for any remaining changes to be made? I've migrated the behaviour completely.

File: airbyte-integrations/connectors/source-twilio/integration_tests/expected_records.jsonl [hunk 5]
```diff
@@ -1,63 +0,0 @@
-{"stream": "addresses", "data": {"sid": "ADa29b1ee20cf61d213f7d7f1a3298309a", "account_sid": "ACdade166c12e160e9ed0a6088226718fb", "friendly_name": null, "customer_name": "test-customer_name_1", "street": "test-street_1", "street_secondary": null, "city": "test-city_1", "region": "test-region_1", "postal_code": "test-postal_code_1", "iso_country": "US", "uri": "/2010-04-01/Accounts/ACdade166c12e160e9ed0a6088226718fb/Addresses/ADa29b1ee20cf61d213f7d7f1a3298309a.json", "date_created": "2020-11-25T09:41:48Z", "date_updated": "2020-11-25T09:41:48Z", "emergency_enabled": false, "validated": false, "verified": false}, "emitted_at": 1682602854052}
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-twilio/integration_tests/expected_records.jsonl:1]: You need to return this data.

**Review by marcosmarxm — CHANGES_REQUESTED** [review]

File: airbyte-integrations/connectors/source-twilio/integration_tests/expected_records.jsonl (hunk 5 above)
Comment (Alcadeus0) [reply airbyte-integrations/connectors/source-twilio/integration_tests/expected_records.jsonl:1]: @marcosmarxm should I be returning the exact data? I've already added the stream to the configured_catalog right with the apporpriate schema. Or is there something else?

Comment (marcosmarxm) [reply airbyte-integrations/connectors/source-twilio/integration_tests/expected_records.jsonl:1]: Yes, this file is needed to run integration tests in our CI env.

Comment (marcosmarxm) [reply airbyte-integrations/connectors/source-twilio/integration_tests/expected_records.jsonl:1]: Also check unit tests are not working too.

Comment (Alcadeus0) [reply airbyte-integrations/connectors/source-twilio/integration_tests/expected_records.jsonl]: @marcosmarxm  okay I have pushed the change to add the records file.
As for the unit tests, how do I import the streams from the manifest.py file into the unit test test_source.py for the test to run? That seems to be the problem from stopping the unit test to be succesful

File: airbyte-integrations/connectors/source-twilio/Dockerfile (hunk 4 above)
Comment (erohmensing) [reply airbyte-integrations/connectors/source-twilio/Dockerfile:37]: @marcosmarxm This is an interesting case. I understand the want to make it breaking if we fear potential breakage (I know we've had issues before where breaking changes are accidentally introduced in the migration due to e.g. catalog changes or spec changes

However if we do things right re: testing and being strict about refactors only being refactors (i.e. if the migration can't be performed without breaking something, maybe it's not a great candidate for low code yet), we wouldn't have to do this. 

I don't think I have a strong opinion either way - use your best judgement :)

**Comment by Alcadeus0** [comment]
@marcosmarxm  Any idea why it fails to pass these tests?

**Comment by RealChrisSean** [comment]
Hello Aazam unfortunately we will not be able to merge this. There are quite a few contributions adding new streams to Twilio making it difficult and more complex than expected :(. We will need to close this.

**Comment by Alcadeus0** [comment]
Hi Chris thanks for notifying. I noticed more streams being merged and understand it would make things much more complex :[
But do we still plan to migrate it to low code sometime in the future? And any plans of hosting another developer hack soon?

**Comment by RealChrisSean** [comment]
We 100% want to migrate it. We have some plan in the works to host a new
hackathon very soon :).

**Pull request closed without merging**